letsencrypt/client/standalone_authenticator.py

```diff
@@ -0,0 +1,519 @@
+        """
+        fork_result = os.fork()
+        Crypto.Random.atfork()
+        if fork_result:
+            # PARENT process (still the Let's Encrypt client process)
+            self.child_pid = fork_result
+            self.do_parent_process(port)
+        else:
+            # CHILD process (the TCP listener subprocess)
+            self.child_pid = os.getpid()
+            self.do_child_process(port, key)
+
+    # IAuthenticator method implementations follow
+
+    def get_chall_pref(self, unused_domain):
+        # pylint: disable=no-self-use
+        """IAuthenticator interface method: Return a list of challenge
+        types that this authenticator can perform for this domain.  In
+        the case of the StandaloneAuthenticator, the only challenge
+        type that can ever be performed is dvsni.
+        """
+        return ["dvsni"]
+
+    def perform(self, chall_list):
+        """IAuthenticator interface method: Attempt to perform the
+        specified challenges, returning the status of each.  For the
+        StandaloneAuthenticator, because there is no convenient way to add
+        additional requests, this should only be invoked once; subsequent
+        invocations are an error.  To perform validations for multiple
+        independent sets of domains, a separate StandaloneAuthenticator
+        should be instantiated.
+        """
+        if self.child_pid or self.tasks:
+            # We should not be willing to continue with perform
+            # if there were existing pending challenges.
+            # TODO: Specify a correct exception subclass.
+            raise Exception(".perform() was called with pending tasks!")
+        results_if_success = []
+        results_if_failure = []
+        if not chall_list or not isinstance(chall_list, list):
+            # TODO: Specify a correct exception subclass.
+            raise Exception(".perform() was called without challenge list")
+        for chall in chall_list:
+            if isinstance(chall, DvsniChall):
+                # We will attempt to do it
+                name, r_b64 = chall.domain, chall.r_b64
+                nonce, key = chall.nonce, chall.key
+                cert, s_b64 = dvsni_gen_cert(name, r_b64, nonce, key)
+                self.tasks[nonce + CONFIG.INVALID_EXT] = cert
+                results_if_success.append({"type": "dvsni", "s": s_b64})
+                results_if_failure.append(None)
+            else:
+                # We will not attempt to do this challenge because it
+                # is not a type we can handle
+                results_if_success.append(False)
+                results_if_failure.append(False)
+        if not self.tasks:
+            # TODO: Specify a correct exception subclass.
+            raise Exception("nothing for .perform() to do")
+        # Try to do the authentication; note that this creates
+        # the listener subprocess via os.fork()
+        if self.start_listener(CONFIG.PORT, key):
```

Comment:
Actually, you found a bug that was introduced when I split out do_parent_process() and do_child_process() from start_listener().  do_parent_process() is capable of returning True or False, and start_listener() needs to return the value that do_parent_process() returns so that it can be checked here.  Originally the return in do_parent_process() existed inside of start_listener() so it was possible for start_listener() to directly return True or False.